lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs

```diff
@@ -843,6 +855,74 @@ prop_runSelectionStep_exceedsTargetAndGetsFurtherAway
+-- Behaviour of selection lenses
+--------------------------------------------------------------------------------
+
+prop_assetSelectionLens_givesPriorityToSingletonAssets
+    :: Blind (Small UTxOIndex)
+    -> Property
+prop_assetSelectionLens_givesPriorityToSingletonAssets (Blind (Small u)) =
+    assetCount >= 2 ==> monadicIO $ do
+        hasSingletonAsset <- isJust <$>
+            run (UTxOIndex.selectRandom u $ WithAssetOnly asset)
+        monitor $ cover 20 hasSingletonAsset
+            "There is at least one singleton entry that matches"
+        monitor $ cover 20 (not hasSingletonAsset)
+            "There are no singleton entries that match"
+        monitor $ counterexample $ unlines
+            ["UTxO index:", pretty $ UTxOIndex.toList u]
+        mUpdatedState <- run $ runSelectionStep lens initialState
+        case mUpdatedState of
+            Nothing -> do
+                -- This should never happen: we should always be able to select
+                -- *something* that matches.
+                monitor $ counterexample "Error: unable to select any entry"
+                assert False
+            Just SelectionState {selected} -> do
+                let output = head $ snd <$> UTxOIndex.toList selected
+                let bundle = view #tokens output
+                case F.toList $ TokenBundle.getAssets bundle of
+                    [a] -> assert $ a == asset
+                    _   -> assert $ not hasSingletonAsset
+  where
+    asset = Set.findMin $ UTxOIndex.assets u
+    assetCount = Set.size $ UTxOIndex.assets u
+    initialState = SelectionState UTxOIndex.empty u
+    lens = assetSelectionLens NoLimit (asset, minimumAssetQuantity)
+    minimumAssetQuantity = TokenQuantity 1
+
+prop_coinSelectionLens_givesPriorityToCoins
+    :: Blind (Small UTxOIndex)
+    -> Property
+prop_coinSelectionLens_givesPriorityToCoins (Blind (Small u)) =
+    entryCount > 0 ==> monadicIO $ do
+        hasCoin <- isJust <$> run (UTxOIndex.selectRandom u WithAdaOnly)
+        monitor $ cover 20 hasCoin
+            "There is at least one coin"
+        monitor $ cover 1 (not hasCoin)
+            "There are no coins"
+        monitor $ counterexample $ unlines
+            ["UTxO index:", pretty $ UTxOIndex.toList u]
+        mUpdatedState <- run $ runSelectionStep lens initialState
+        case mUpdatedState of
+            Nothing -> do
+                -- This should never happen: we should always be able to select
+                -- *something* that matches.
+                monitor $ counterexample "Error: unable to select any entry"
+                assert False
+            Just SelectionState {selected} -> do
+                let output = head $ snd <$> UTxOIndex.toList selected
+                let bundle = view #tokens output
+                case F.toList $ TokenBundle.getAssets bundle of
+                    [] -> assert hasCoin
+                    _  -> assert $ not hasCoin
```

Comment:
:+1: , essentially the same as above with a different selector.

Reply:
Yes, indeed!

There might be a nice way to extract out the common functionality into a separate function. Though I guess we could leave that to a future PR. 😅